### ISSUE TITLE
[mimir-distributed] Consider using own memcached manifest instead of bitnami chart?

### ISSUE THREAD

**Issue by Skaronator** [grafana/helm-charts#1315]
Currently, the mimir-distributed helm-chart uses the bitnami memcached helm-chart instead of its own manifest files like the loki-distributed chart. The Bitnami chart doesn't work out-of-the-box on OpenShift/OKD due to SecurityContextContrains.

Installing mimir-distributed with memcached enabled results in the following error:

> FailedCreate  statefulset-controller  create Pod mimir-distributed-memcached-0 in StatefulSet mimir-distributed-memcached failed error: pods "mimir-distributed-memcached-0" is forbidden: unable to validate against any security context constraint: [provider "scc-custom-order-service": Forbidden: not usable by user or serviceaccount, provider "anyuid": Forbidden: not usable by user or serviceaccount, provider "scc-custom-default": Forbidden: not usable by user or serviceaccount, provider "uba-default": Forbidden: not usable by user or serviceaccount, provider "vault": Forbidden: not usable by user or serviceaccount, provider "uba-anti-virus-service": Forbidden: not usable by user or serviceaccount, provider "custom-postgres": Forbidden: not usable by user or serviceaccount, provider restricted: .spec.securityContext.fsGroup: Invalid value: [[]int64{1001}: 1001 is not an allowed group, spec.containers[0].securityContext.runAsUser: Invalid value: 1001: must be in the ranges: [1000660000, 1000669999], spec.containers[1].securityContext.runAsUser: Invalid value: 1001: must be in the ranges: [1000660000, 1000669999], provider "nonroot": Forbidden: not usable by user or serviceaccount, provider "loki-distributed": Forbidden: not usable by user or serviceaccount, provider "hostmount-anyuid": Forbidden: not usable by user or serviceaccount, provider "machine-api-termination-handler": Forbidden: not usable by user or serviceaccount, provider "hostnetwork": Forbidden: not usable by user or serviceaccount, provider "hostaccess": Forbidden: not usable by user or serviceaccount, provider "node-exporter": Forbidden: not usable by user or serviceaccount, provider "privileged": Forbidden: not usable by user or serviceaccount, provider "scc-custom-storage": Forbidden: not usable by user or serviceaccount]

The mimir-distributed chart includes a PSP, which would _probably_ fix it, but that doesn't apply to sub-charts since they have their own SA/Role.
https://github.com/grafana/helm-charts/blob/f1bc352d5272ec5e5e50665e9ac720c66fe9f252/charts/mimir-distributed/templates/podsecuritypolicy.yaml#L20
https://github.com/grafana/helm-charts/blob/f1bc352d5272ec5e5e50665e9ac720c66fe9f252/charts/mimir-distributed/templates/podsecuritypolicy.yaml#L31-L32

Additionally, the bitnami chart provides some high-availablity mode and persistence, which confused me when I first saw that in the values.yaml. It seems overkill to me and doesn't work either, since the mimir-distributed chart overwrites the bitnami image with the offical image:
https://github.com/grafana/helm-charts/blob/f1bc352d5272ec5e5e50665e9ac720c66fe9f252/charts/mimir-distributed/values.yaml#L969
https://github.com/grafana/helm-charts/blob/f1bc352d5272ec5e5e50665e9ac720c66fe9f252/charts/mimir-distributed/values.yaml#L977-L979

Debugging the bitnami chart wasn't easy either, since I didn't notice that mimir-distributed still uses the older 5.x version instead of the 6.x release from bitnami:
https://github.com/grafana/helm-charts/blob/f1bc352d5272ec5e5e50665e9ac720c66fe9f252/charts/mimir-distributed/Chart.yaml#L10-L13

My wish would be that mimir-distributed uses the same approach as loki-distributed when it comes to memcached. I would also recommend renaming the "memcached" that doesn't have a purposeful name to something better.

https://github.com/grafana/helm-charts/blob/f1bc352d5272ec5e5e50665e9ac720c66fe9f252/charts/mimir-distributed/Chart.yaml#L11-L12

**Comment by krajorama** [comment]
Hi, related to #1217 . We (cc @patrickabrennan) have it running in openshift now with a newer bitnami chart that can have it's ServiceAccount set. We still need to find the time to make a proper PR out of it though.

https://github.com/grafana/helm-charts/tree/krajo/mimir-update-memcached + needs Patrick's SCC stuff which I'll add.

As much as I'd like to align with the loki chart, I'd also prefer to not have to maintain memcached stuff ourselves. If we have the proper support for openshift with a recent bitnami chart, would that be acceptable for your?

**Comment by Skaronator** [comment]
Not having to maintain memcached stuff is a valid argument. I'm not the biggest fan of the Bitnami stuff, since they have a lot of magic in their images (which was ultimately replaced by the official container image in this case). I'm not a maintainer of that stuff, so it is still your decision at the end of the day. :)

The following things would still be great to archive:
- Upgrade Bitnami memcached chart to 6.x
- Give the [normal memcached](https://github.com/grafana/helm-charts/blob/f1bc352d5272ec5e5e50665e9ac720c66fe9f252/charts/mimir-distributed/Chart.yaml#L11-L12) a suffix, what it's actually doing. Maybe "memcached-chunks" would match the best, it's similar to the loki-distributed chart :)
- Update and use the alpine version of the memcached container image, similar to loki-distributed

**Comment by krajorama** [comment]
Let me talk to my colleague Patrick and come to some decision . I'm not sure we can do the suffix without breaking change, but everything else is in scope. When we have the PR going would you be open to beta test it ?

**Comment by Skaronator** [comment]
Yeah, I can give it a try. Would be good if you can provide the hosted artifact of that helm chart, since I have to deploy it with the CI and my local kubernetes cluster isn't an OKD.

The breaking change is not mission-critical, maybe something for #1309.

**Comment by krajorama** [comment]
I gave it a go, but the bitnami chart has a hundred settings and it makes no sense to replicate those (and probably introduce errors). So switching to our own implementation has to wait until #1309. We're going to update my branch to be mergeable for short term openshift support (https://github.com/grafana/helm-charts/commits/krajo/mimir-update-memcached).

We've already checked the changelogs and we don't see any compatibility issue when upgrading to a fairly new bitnami chart.

**Comment by krajorama** [comment]
Hi, added a draft PR that we tried in one of our test systems. Looking into providing artifact.

**Comment by krajorama** [comment]
Hi, looking into your question (my expert is in different time zone). Also I got no better idea for sharing the chart than creating a copy. Will call it mimir-openshift-experimental . (Unless it's unusable due to some naming limit, going to test).

**Comment by krajorama** [comment]
Released the experimental chart: 
grafana/mimir-openshift-experimental	2.1.0        	2.0.0      	Grafana Mimir on OpenShift Experiment

I tested it before the release locally, but no time yet to check that the release actually worked :(

**Comment by Skaronator** [comment]
Hi, just gave this a test and everything seems to work just fine!

**Comment by krajorama** [comment]
We've removed the dependency on the bitnami chart and moved to using own manifests.
